### PR TITLE
Email verification

### PR DESCRIPTION
**Pull request by aMarshmallows**
.eslintrc.js was changed for some reason so I ran into a lot of lint errors. if someone can help, I put that back to auto in the first commit in this PR.

### CODE REVIEW

**Comment by rob-3** [comment]
One thing you can do first is fix the merge conflict!

**Comment by Sleepingpirates** [comment]
Jwt would be better for email verification instead of just userid

**Comment by rob-3** [comment]
> Jwt would be better for email verification instead of just userid

Yes, as it stands rn this isn't the most robust solution because the userid is not unguessable - ie someone could just guess it and visit the link without actually getting the email.

**Comment by Sleepingpirates** [comment]
> > Jwt would be better for email verification instead of just userid
> 
> Yes, as it stands rn this isn't the most robust solution because the userid is not unguessable - ie someone could just guess it and visit the link without actually getting the email.

They can just get the user Id from local storage lol

**Comment by rob-3** [comment]
> They can just get the user Id from local storage lol

Oh true. Yeah that makes this very easily bypassable

File: web/src/pages/api/user/register.ts [hunk 1]
```diff
@@ -45,7 +47,7 @@ export default async function Register(req: Request, res: NextApiResponse) {
         },
       });
 
-    const confirmEmailAPIRoute = `http://localhost:3000/api/user/confirmEmail/${newUser.id}`;
+    const confirmEmailAPIRoute = `http://localhost:3000/api/user/confirmEmail/id?=${newUser.id}token?=${jwt}`;
```
Comment (Rikimbili) [inline web/src/pages/api/user/register.ts:64]: We should add a check here to assign the right hostname. Also the slug seems to be wrong. Here is what I believe is the correct code with the environment check:
```
  const url =
    process.env.NODE_ENV === "development"
      ? "http://localhost:3000"
      : "https://daydream.wtf";
  const confirmEmailAPIRoute = `${url}/api/user/confirmEmail?userId=${newUser.id}&token=${jwt}`;
```

File: web/src/utils/jwt.ts [hunk 2]
```diff
@@ -25,6 +25,29 @@ export const generateJWT = async (userId: string): Promise<string> => {
...
 };
 
+export const generateJWTForEmailVerification = async (
+  userId: string
+): Promise<string> => {
+  return new Promise((resolve, reject) => {
+    sign(
+      { userId },
+      env.JWT_PRIVATE_KEY,
+      { expiresIn: "365d" },
```
Comment (Rikimbili) [inline web/src/utils/jwt.ts:35]: We might want to make this expire in 20 mins. What do you think @rob-3 @Sleepingpirates ?

File: web/src/pages/api/user/confirmEmail.ts [hunk 3]
```diff
@@ -1,23 +1,35 @@
...
+import { validateMethod } from "../../../utils/validation";
 
 export default async function confirmEmail(
   req: NextApiRequest,
   res: NextApiResponse
 ) {
   if (!validateMethod("GET", req, res)) return;
 
-  const { userID } = req.query;
+  const userID = req.query.id;
```
Comment (Rikimbili) [inline web/src/pages/api/user/confirmEmail.ts:14]: To keep it consistent the other endpoints let's name it `userId` with lowercase d

File: web/src/pages/api/user/confirmEmail.ts [hunk 4]
```diff
@@ -1,23 +1,35 @@
...
+  const token = req.query.token;
 
-  if (Array.isArray(userID) || userID === undefined) {
+  if (
+    Array.isArray(userID) ||
+    userID === undefined ||
+    Array.isArray(token) ||
+    token === undefined
+  ) {
```
Comment (Rikimbili) [inline web/src/pages/api/user/confirmEmail.ts:22]: This can be reduced to 

`if (!userId || !token || Array.isArray(userId) || Array.isArray(token))`

**Review by Rikimbili — CHANGES_REQUESTED** [review]

File: web/src/pages/api/user/confirmEmail.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  });
+
+  // if user does exist, then create a date object and add that user's emailVerified date
+  if (maybeUser) {
+    await prisma.user.update({
+      where: {
+        id: maybeUser.id,
+      },
+      data: {
+        emailVerified: new Date(),
```
Comment (rob-3) [inline web/src/pages/api/user/confirmEmail.ts:47]: Have you verified that this actually works? I don't know if prisma can handle direct dates like that

Comment (aMarshmallows) [reply web/src/pages/api/user/confirmEmail.ts:47]: This is what I saw in the Prisma docs so I assumed it was correct but I don't know another way of doing it hmm

**Review by Rikimbili — APPROVED** [review]

**Review by rob-3 — APPROVED** [review]

**Pull request merged**